File: services/firecrawl_service.py
from firecrawl import FirecrawlApp
from typing import Dict, List, Any, Optional
import logging

logger = logging.getLogger(__name__)

class FirecrawlService:
    """Service for web scraping using Firecrawl API"""
    
    def __init__(self, api_key: str):
        if not api_key:
            raise ValueError("Firecrawl API key is required")
        self.app = FirecrawlApp(api_key=api_key)
    
    def _extract_dict_data(self, obj: Any) -> Dict[str, Any]:
        """
        Helper method to extract dictionary data from various object types
        including Pydantic models (v1 and v2)
        """
        if isinstance(obj, dict):
            return obj
        elif hasattr(obj, 'model_dump'):  # Pydantic v2
            return obj.model_dump()
        elif hasattr(obj, 'dict'):  # Pydantic v1
            return obj.dict()
        elif hasattr(obj, '__dict__'):  # Regular object
            return obj.__dict__
        else:
            return {}
    
    def scrape_url(self, url: str) -> Dict[str, Any]:
        """
        Scrape a single URL and return the content
        
        Args:
            url: The URL to scrape
            
        Returns:
            Dict containing url, markdown, title, and description
        """
        try:
            result = self.app.scrape_url(
                url,
                formats=['markdown'],
                only_main_content=True
            )
            
            if not result:
                raise Exception('Failed to scrape content from the URL')
            
            # Convert result to dictionary safely
            result_dict = self._extract_dict_data(result)
            
            markdown = result_dict.get('markdown', '')
            metadata = result_dict.get('metadata', {})
            
            if not markdown:
                raise Exception('No markdown content returned from the URL')
            
            # Convert metadata to dictionary safely
            metadata_dict = self._extract_dict_data(metadata)
            
            return {
                'url': metadata_dict.get('sourceURL', url),
                'markdown': markdown,
                'title': metadata_dict.get('title'),
                'description': metadata_dict.get('description')
            }
            
        except Exception as e:
            logger.error(f"Error scraping URL {url}: {str(e)}")
            raise Exception(f"Failed to scrape URL: {str(e)}") 

File: services/firecrawlService.ts
import FirecrawlApp from '@mendable/firecrawl-js';

interface CrawlResult {
  url: string;
  markdown: string;
  title?: string;
  description?: string;
}

class FirecrawlService {
  private app: FirecrawlApp;

  constructor(apiKey?: string) {
    if (!apiKey) {
      throw new Error('Firecrawl API key is required. Please set your API key.');
    }
    this.app = new FirecrawlApp({ apiKey });
  }

  async scrapeUrl(url: string): Promise<CrawlResult> {
    try {
      const result = await this.app.scrapeUrl(url, {
        formats: ['markdown'],
        onlyMainContent: true
      });

      if (!result || !result.markdown) {
        throw new Error('Failed to scrape content from the URL');
      }

      return {
        url: result.url || url,
        markdown: result.markdown,
        title: result.metadata?.title,
        description: result.metadata?.description
      };
    } catch (error: any) {
      console.error('Error scraping URL:', error);
      throw new Error(`Failed to crawl URL: ${error.message || 'Unknown error'}`);
    }
  }

  async crawlWebsite(url: string, maxPages: number = 10): Promise<CrawlResult[]> {
    try {
      const crawlResult = await this.app.crawlUrl(url, {
        limit: maxPages,
        scrapeOptions: {
          formats: ['markdown'],
          onlyMainContent: true
        }
      });

      if (!crawlResult || !crawlResult.data) {
        throw new Error('Failed to crawl website');
      }

      return crawlResult.data.map((item: any) => ({
        url: item.url,
        markdown: item.markdown || '',
        title: item.metadata?.title,
        description: item.metadata?.description
      }));
    } catch (error: any) {
      console.error('Error crawling website:', error);
      throw new Error(`Failed to crawl website: ${error.message || 'Unknown error'}`);
    }
  }
}

export { FirecrawlService, type CrawlResult }; 